appengine-jetty-managed-runtime/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeWebAppContext.java

```diff
@@ -262,129 +278,122 @@ public boolean isTrustedRemoteAddr(String remoteAddr) {
     }
 
     @Override
-    public String getServerInfo() {
-      return serverInfo;
-    }
+    public void requestInitialized(ServletRequestEvent sre) {
+      ServletRequest request = sre.getServletRequest();
+      Request baseRequest = Request.getBaseRequest(request);
+      RequestContext requestContext = getRequestContext(baseRequest);
+      
+      // Check for SkipAdminCheck and set attributes accordingly.
+      VmRuntimeUtils.handleSkipAdminCheck(requestContext);
 
+      // Change scheme to HTTPS based on headers set by the appserver.
+      setSchemeAndPort(baseRequest);
+    }
+    
     @Override
-    public void log(String message) {
-      log(message, null);
+    public void requestDestroyed(ServletRequestEvent sre) {      
+      ServletRequest request = sre.getServletRequest();
+      Request baseRequest = Request.getBaseRequest(request);
+      
+      if (request.isAsyncStarted()) {
+        request.getAsyncContext().addListener(new AsyncListener() {
+          @Override public void onTimeout(AsyncEvent event) throws IOException {}
+          @Override public void onStartAsync(AsyncEvent event) throws IOException {}
+          @Override public void onError(AsyncEvent event) throws IOException {}
+          
+          @Override
+          public void onComplete(AsyncEvent event) throws IOException {
+            complete(baseRequest);
+          }
+        });
+      } else {         
+        complete(baseRequest);  
+      }
     }
-
-    /**
-     * {@inheritDoc}
-     *
-     * @param throwable an exception associated with this log message,
-     * or {@code null}.
-     */
+    
     @Override
-    public void log(String message, Throwable throwable) {
-      StringWriter writer = new StringWriter();
-      writer.append("javax.servlet.ServletContext log: ");
-      writer.append(message);
-
-      if (throwable != null) {
-        writer.append("\n");
-        throwable.printStackTrace(new PrintWriter(writer));
+    public void exitScope(org.eclipse.jetty.server.handler.ContextHandler.Context context, Request baseRequest) {
+      if (LOG.isDebugEnabled())
+        LOG.debug("Exit {} -> {}",ApiProxy.getCurrentEnvironment(),defaultEnvironment);
+      ApiProxy.setEnvironmentForCurrentThread(defaultEnvironment);
+    }
+    
+    private void complete(Request baseRequest)
+    {
+      RequestContext requestContext = getRequestContext(baseRequest);
+      VmApiProxyEnvironment env = requestContext.getRequestSpecificEnvironment();
+
+      // Transaction Cleanup 
+      Collection<Transaction> txns = datastoreService.getActiveTransactions();
+      if (!txns.isEmpty()) {
+        handleAbandonedTxns(txns);
+      }
+     
+      // Save dirty sessions
+      HttpSession session = baseRequest.getSession(false);
+      if (session instanceof AppEngineSession) {
+        final AppEngineSession aeSession = (AppEngineSession) session;
+              if (aeSession.isDirty())
+                aeSession.save();
       }
```

Comment:
@ludoch 
I've moved the behaviour of SaveServletFilter and AbandonedTransactionFilter to the complete method above.  This avoids the need for 3 async listeners, each with an onComplete handling.   Behaviour should be the same.